draft-ietf-quic-recovery.md

```diff
@@ -1301,9 +1303,8 @@ SetLossDetectionTimer():
     assert(!PeerCompletedAddressValidation())
     sent_time = now()
 
-  // Calculate PTO duration
-  timeout = smoothed_rtt + max(4 * rttvar, kGranularity) +
-    max_ack_delay
+  // Calculate PTO duration, sent_time includes max_ack_delay.
+  timeout = smoothed_rtt + max(4 * rttvar, kGranularity)
```

Comment:
Not that it was great before, but I think that this change obfuscates things.

I think that the inclusion of the max_ack_delay argument overloads GetEarliestTimeAndSpace in a way that just highlights how the dual purpose to which the function is applied is not appropriate.

We use this function for working out two things:
* when to declare something lost
* when the PTO timer should start 

The former is not affected by max_ack_delay.   The PTO calculation is, but this change adds max_ack_delay to the baseline rather than the PTO calculation itself.

For one, that's a fundamental change in logic for one (because PTO number 2 used to include 2*max_ack_delay and now it doesn't. Maybe that was a bug that this fixes, but that needs to be clear.

However, the main problem is the way that this code now communicates its intent.  If the intent is to start with a baseline (t_sent) and to add a PTO (RTT + 4*rttvar) that is then doubled each time, and only then compensate for ACK delay, you want to show your work, not hide it:

```
pto_timer = smoothed_rtt + max(4 * rttvar, kGranularity)
pto_timer *= 1 << pto_count
loss_detection_timer.update(sent_time + pto_timer + ?max_ack_delay)
```
Or is it, with maybe a fix for the packet number space issue...
```
pto_timer = smoothed_rtt + max(4 * rttvar, kGranularity) + ?max_ack_delay
pto_timer *= 1 << pto_count
loss_detection_timer.update(sent_time + pto_timer)
```

The trick is that the code that triggers SendOneOrTwoAckElicitingPackets() needs to follow the same logic.  But that too is too clever.  Basing that on the reference time (the time you sent the last packet) rather than the time that the timer is expected to pop is obtuse (even if it is correct as you currently formulate it).

Why not provide a CalculatePtoTime() function and use that?  This code doesn't need to be optimized, so you can calculate everything always.

```python
pto_time = { Initial: 0, Handshake: 0, ApplicationData: 0 };
for space in time_of_last_sent_ack_eliciting_packet:
    if time_of_last_sent_ack_eliciting_packet[space] == 0:
        continue;
    t = smoothed_rtt + max(4 * rttvar, kGranularity)
    t = t * (2 ^ pto_count)
    if space == ApplicationData: # or move this up as you need
        t += max_ack_delay
    pto_time[space] = time_of_last_sent_ack_eliciting_packet[space] + t
pto_time, pn_space = GetEarliestTimeAndSpace(pto_time)
```

This needs tweaking for the case where sent_time == 0, but that is just a case of factoring out the PTO interval calculation.  Note that you aren't adding max_ack_delay in that case, which is another difference.